pkg/storage/unified/resource/search.go

```diff
@@ -768,30 +799,105 @@ func (s *searchSupport) build(ctx context.Context, nsr NamespacedResource, size
 				// When we reach the batch size, perform bulk index and reset the batch.
 				if len(items) >= maxBatchSize {
 					span.AddEvent("bulk indexing", trace.WithAttributes(attribute.Int("count", len(items))))
-					if err = index.BulkIndex(&BulkIndexRequest{
-						Items: items,
-					}); err != nil {
+					if err = index.BulkIndex(&BulkIndexRequest{Items: items}); err != nil {
 						return err
 					}
 
-					// Reset the slice for the next batch while preserving capacity.
 					items = items[:0]
 				}
 			}
 
 			// Index any remaining items in the final batch.
 			if len(items) > 0 {
 				span.AddEvent("bulk indexing", trace.WithAttributes(attribute.Int("count", len(items))))
-				if err = index.BulkIndex(&BulkIndexRequest{
-					Items: items,
-				}); err != nil {
+				if err = index.BulkIndex(&BulkIndexRequest{Items: items}); err != nil {
 					return err
 				}
 			}
 			return iter.Error()
 		})
 		return listRV, err
-	})
+	}
+
+	updaterFn := func(context context.Context, index ResourceIndex, sinceRV int64) (int64, int, error) {
+		span := trace.SpanFromContext(ctx)
+		span.AddEvent("updating index", trace.WithAttributes(attribute.Int64("sinceRV", documentStatsRV)))
+
+		rv, it := s.storage.ListModifiedSince(ctx, NamespacedResource{
+			Group:     nsr.Group,
+			Resource:  nsr.Resource,
+			Namespace: nsr.Namespace,
+		}, sinceRV)
+
+		// Process documents in batches to avoid memory issues
+		// When dealing with large collections (e.g., 100k+ documents),
+		// loading all documents into memory at once can cause OOM errors.
+		items := make([]*BulkIndexItem, 0, maxBatchSize)
+
+		docs := 0
+		for res, err := range it {
+			// Finish quickly if context is done.
+			if ctx.Err() != nil {
+				return 0, 0, ctx.Err()
+			}
+
+			docs++
+
+			if err != nil {
+				span.RecordError(err)
+				return 0, 0, err
+			}
+
+			key := &res.Key
+			switch res.Action {
+			case resourcepb.WatchEvent_ADDED, resourcepb.WatchEvent_MODIFIED:
+				span.AddEvent("building document", trace.WithAttributes(attribute.String("name", res.Key.Name)))
+				// Convert it to an indexable document
+				doc, err := builder.BuildDocument(ctx, key, res.ResourceVersion, res.Value)
+				if err != nil {
+					span.RecordError(err)
+					logger.Error("error building search document", "key", SearchID(key), "err", err)
+					continue
+				}
+
+				items = append(items, &BulkIndexItem{
+					Action: ActionIndex,
+					Doc:    doc,
+				})
+			case resourcepb.WatchEvent_DELETED:
+				span.AddEvent("deleting document", trace.WithAttributes(attribute.String("name", res.Key.Name)))
+				items = append(items, &BulkIndexItem{
+					Action: ActionDelete,
+					Key:    &res.Key,
+				})
+			default:
+				logger.Error("can't update index with item, unknown action", "action", res.Action, "key", key)
```

Comment:
I suppose in the query for `ListModifiedSince()` I should have added `WHERE action IN (1,2,3)`. Can add that later though 👍